### PR TITLE
TFM mappings for WebAssembly and MonoUE

### PR DESCRIPTION
**Pull request by mhutch**
Added TFMs for WebAssembly and Mono for Unreal Engine.

Like other Mono-based frameworks, these are compatible with netstandard 2.0.

### CODE REVIEW

**Comment by rohit21agrawal** [comment]
@joelverhagen you know this area, could you take a peek please?

File: src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: Is this `WebAssembly` TFM really appropriate? What really defines the compatibility level between WebAssembly and .NET Standard 2.0? Said another way, is there an official compatibility matrix between WebAssembly and .NET Standard assemblies? Truth be told, I am pretty ignorant about .NET + WebAssembly (and WebAssembly in general). This is the closest thing I could find: https://github.com/xamarin/WebSharp.

At first glance, this seems overly generic and perhaps could be renamed to the tool that is interopping between .NET Standard assemblies and WebAssembly.

File: test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs
Comment (joelverhagen) [inline test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:907]: Can we have some tests asserting that `wasm` and `monoue` does not support .NET Standard 2.1? Essentially a test verifying the max .NET Standard version is exactly what is expected.

**Review by joelverhagen — DISMISSED** [review]
⌚️

File: src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs
Comment (joelverhagen) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: To be clear I am not totally pushing back on the name, just trying to understand the full picture.

File: test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs
Comment (mhutch) [reply test/NuGet.Core.Tests/NuGet.Frameworks.Test/CompatibilityTests.cs:907]: We don't do this for any of the other frameworks, and I suspect them to support 2.1 pretty quickly.

I'd love to delegate this information to the framework instead of centralizing it all in one place.

File: src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs
Comment (mhutch) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: It's [a new framework based on the Mono runtime](http://www.mono-project.com/news/2017/08/09/hello-webassembly/) that runs in the browser and has access to browser APIs. See also https://github.com/mono/mono/tree/master/sdks/wasm. This is the framework used by [Blazor](https://github.com/aspnet/Blazor).

netstandard compat level is the same as the Xamarin frameworks, i.e. unversioned and assumed to be latest.

cc @migueldeicaza @terrajobst

Comment (marek-safar) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: It might be more appropriate to name it MonoWebAssembly or MonoWasm

**Comment by rohit21agrawal** [comment]
@mhutch if you still want this PR to go in, can you please address the feedback?

File: src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs
Comment (migueldeicaza) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: The WebAssembly target is compatible with .NET Standard 2, but it is a superset of that API, as it is provides many capabilities not found on .NET Standard 2, it is based on the various mobile profiles used by Xamarin.

I do not think that we need to brand this as a Mono web assembly, it is not an implementation specific profile.

**Comment by mhutch** [comment]
@rohit21agrawal re. the TFM name, see Miguel's response.

Re. the request for tests, none of the existing TFMs have the suggested tests, and I don't think wasm is any different than them.

**Comment by danroth27** [comment]
@davidfowl

**Comment by danroth27** [comment]
One concern with having a generic WebAssembly TFM is how would you reason about targeting WebAssembly in browsers vs running in other environments that support WebAssembly but don't have the same limitations? For example, what if you are running on WebAssembly in Node or in an Electron app?

**Comment by terrajobst** [comment]
Two questions:

1. Which API surface/platform does `wasm` represent?
2. What's the mapping between .NET Standard and `wasm`?

File: src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs
Comment (terrajobst) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: > it is not an implementation specific profile.

Then what is it? The only implementation free TFM we have is `netstandard` and this is mapped to every TFM. `wasm` wouldn't be mapped to anything but a specific implementation, so I don't understand how this isn't inherently an implementation specific profile.

**Comment by marek-safar** [comment]
@terrajobst 

1. The API surface will probably closest to the latest netcore
2. What do you mean by mapping? It'll be netstandard 2 compatible if that's what you are asking

**Comment by jeromelaban** [comment]
@marek-safar it will be close to netstandard2.0, but it will have web assembly specific apis (https://github.com/mono/mono/blob/8de852ea9e611c3e841d33ad26dde8bd034e4325/sdks/wasm/bindings.cs) so it should be "mono5.x-wasm1.0" ?

**Comment by marek-safar** [comment]
That's fine, we'll just expose this API as a different package. It does not affect how netstandard is defined. Actually, the exact opposite is needed. We need to control how the netstandard implementation works on WASM target. There was discussion above where I suggested having more fine-grained control by defining Mono specific TFM so the netstandard packages could have in theory multiple WASM backends support, similar to coreclr/mono support but it looks like folks are in favour of single TFM.

**Comment by jeromelaban** [comment]
Sure, it's in a different package, but what's the TFM for this one? If wasm replicates the platform vs runtime ambiguity for Xamarin.ios and MonoAndroid, this is still compatibility issue regarding the Platform/BCL/Runtime version matrix.

wasm1.0 is a platform, and if the definition stays this way, it will implicitly mean wasm1.0 running on mono runtime because of internal calls dependencies. What if .NET core happens to support wasm as well ?

**Comment by gulbanana** [comment]
any progress on choosing a tfm name? as a library developer, i cannot support blazor until cross-targeting is available. the runtime environment is too different (no threads etc)

**Comment by kzu** [comment]
Would love to see this merged. I don't think there are outstanding issues, are there?

**Comment by rohit21agrawal** [comment]
the test failures on this PR are an issue for now, can @mhutch also rebase and push so we can validate against the latest dev branch?

**Comment by mhutch** [comment]
@rohit21agrawal rebased and pushed

**Comment by mhutch** [comment]
@rohit21agrawal looks like it failed due to the VSTS outage. Can we re-run it?

**Comment by mhutch** [comment]
It doesn't look like the failing tests are related to these changes. What can I do to move this along?

**Comment by rohit21agrawal** [comment]
the mac test failures are definitely related: https://devdiv.visualstudio.com/DevDiv/NuGet/_TestManagement/Runs?runId=2303373&_a=resultQueryEditor

**Comment by mhutch** [comment]
Sorry, didn't see those among all the random failures. Fixed!

**Comment by rohit21agrawal** [comment]
@mhutch CompatibilityListProvider tests are still failing on linux/mac/windows . You can ignore the functional test failures related to UwpBlankApp and Apex Test Failures

**Comment by mhutch** [comment]
ok, i think i fixed it.

File: src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: @migueldeicaza @mhutch can you please help alleviate @terrajobst 's concern before we merge this?

Comment (SteveSandersonMS) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: If this TFM was called `MonoWebAssembly`, and then in the future other WebAssembly implementations appear that share some commonality, would it be possible to introduce an intermediate non-implementation-specific `WebAssembly` TFM at that time (so the `MonoWebAssembly` and `OtherWebAssembly` would both become children of `WebAssembly`)?

Comment (jeromelaban) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: A `WebAssembly` TFM would make sense if there were a stable subset of APIs specific to WebAssembly, a `WebAssemblyStandard 1.0`, such as this:

https://github.com/mono/mono/blob/master/sdks/wasm/bindings.cs#L21

But are we already in such a state to know about this ? Currently, I'm blocked in many occasions by the fact that making nuget packages is not possible, and also because I cannot reference a package that provides this Runtime class.

Similarly, should there be a `WebAssemblyApp` and `MonoWebAssemblyApp` ?

Even if there is only `MonoWebAssembly` or `NetCoreWebAssembly` to begin with, re-targetting to `WebAssembly` when it will make sense is now very easy, and could be dealt with the _runtime_ nuget feature.

**Comment by mhutch** [comment]
I have some packages that need to use this TFM :/

File: src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs
Comment (terrajobst) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: I agree with @jeromelaban that an abstract `WebAssembly` TFM would be like .NET Standard. Which also explains why I don't think we should have another one -- we started .NET Standard specifically to unify all existing portability mechanisms (Silverlight, PCL, contracts) into a single thing that was simpler & more viable (i.e. has enough APIs). Moving forward, we don't want specific TFM mappings anymore; rather an asset should only be considered compatible if it matches the project's TFM or the corresponding .NET Standard version.

Personally, I think we need a TFM that designates a particular implementation of web assembly, rather than the concept. And `MonoWebAssembly` sounds like a fine plan to me.

**Comment by jainaashish** [comment]
@mhutch Without understanding the background and proper scenario for this PR and analyzing alternate means to achieve it, we won't be able to proceed. So I'd suggest to set some time with @rrelyea to discuss and proceed with the PR.

File: src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs
Comment (gulbanana) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: while MonoWebAssembly would be fine in practice, i want to note that as a library author the reasons i need a tfm here are not mono-specific. i need to make multitargeted builds which do *not* access certain features of .net standard (threading, filesystem) and instead use less efficient implementations or provide fewer features. these are limitations of wasm conceptually, not limitations of mono’s wasm implementation.

this doesn’t matter much since a hypothetical future with `<TargetFrameworks>monowasm;corewasm<TargetFrameworks>` would not be very costly

Comment (danroth27) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: > i need to make multitargeted builds which do not access certain features of .net standard (threading, filesystem) and instead use less efficient implementations or provide fewer features
> these are limitations of wasm conceptually, not limitations of mono’s wasm implementation

I believe these are more the limitations of the browser then of WebAssembly itself. For example, I believe you could write a different WebAssembly host (perhaps something like [wasmjit](https://github.com/rianhunter/wasmjit)) that doesn't have these restrictions.

Comment (gulbanana) [reply src/NuGet.Core/NuGet.Frameworks/FrameworkConstants.cs:71]: yes, i suppose it's a matter of what 'syscalls' (or DOM bindings) are exposed to the WASM host

**Comment by jainaashish** [comment]
There is no point in keeping this PR open at this time, until this has been discussed and finalized. So I'm closing this PR for now.

**Comment by jeromelaban** [comment]
@jainaashish is there a opened issue for discussing this ?

**Comment by jainaashish** [comment]
Not that I see any.. @mhutch can you open a workitem with all the details?

**Comment by gulbanana** [comment]
The lack of a tfm other than netstandard is still blocking me from porting libraries to Blazor.

**Comment by weitzhandler** [comment]
Ping.
This issue is blocking authors of .NET Standard frameworks to have wasm-specific ports. Can you please dedicate a TFM?

**Comment by weitzhandler** [comment]
Ping.
This issue is blocking authors of .NET Standard frameworks to have wasm-specific ports. Can you please dedicate a TFM?
Could this issue be reopened?

@jainaashish have you had the chance to discuss this with @rrelyea?

**Comment by jainaashish** [comment]
@nkolev92 @anangaur

**Comment by nkolev92** [comment]
@weitzhandler 
These are decisions that are going to be made with other TFM stakeholders (not unilaterally by NuGet)
Please continue the discussion on the following issue: https://github.com/NuGet/Home/issues/8186

**Comment by weitzhandler** [comment]
@nkolev92 I wish there was `WebAssembly` even as a general TFM temporarily, or any other immediate solution. Right now, NuGet and wasm don't work well together, can't build platform wasm projects with NuGet.
This is like a blocking issue.

**Comment by nkolev92** [comment]
@weitzhandler 

TFM changes are never temporary though. 

If/when a new tfm is added, the scenario should be well understood and in conformance to where we think the eco-system needs to move, not as a workaround. 

Regardless, expect updates on https://github.com/NuGet/Home/issues/8186.

**Comment by weitzhandler** [comment]
Then a general wasm should be added permanently, then when we're ready, all the various dialects.

**Comment by nkolev92** [comment]
I think the argument on the other issue is that we should not keep adding different dialects through additional frameworks. 
Having too many frameworks was not a great experience in the past and we're trying to avoid it here. 

The folks on that thread(Immo, Rich) have better context and know more about this topic than I ever will so I will defer to them. :)

**Comment by weitzhandler** [comment]
Working with wasm has become a dead-end with no TFM for it.
I see your point tho. Thanks.

**Pull request closed without merging**
